src/uTensor/allocators/arenaAllocator.hpp

```diff
@@ -0,0 +1,372 @@
+#include <algorithm>  // for sort function
+#include <cstdio>
+#include <forward_list>
+#include <memory>
+#include <vector>
+
+#include "context.hpp"
+#include "memoryManagementInterface.hpp"
+#include "tensor.hpp"
+
+namespace uTensor {
+
+//#define MSB_SET ~( ~( (T)0 ) >> 1 )
+#define MSB_SET (1 << (sizeof(T) * 8 - 1))
+#define BLOCK_INACTIVE ~MSB_SET
+#define BLOCK_LENGTH_MASK ~MSB_SET
+#define BLOCK_ACTIVE MSB_SET
+#define BLOCK_ZERO_LENGTH 0
+
+/** EVENT LIST
+ */
+
+DECLARE_EVENT(MetaHeaderNotFound);
+DECLARE_EVENT(localCircularArenaAllocatorRebalancing);
+DECLARE_EVENT(localCircularArenaAllocatorConstructed);
+DECLARE_ERROR(InvalidBoundRegionState);
+
+template <typename T>
+constexpr size_t meta_addressable_space() {
+  return 1 << (sizeof(T) * 8 - 1);
+}
+/**
+ * Size allocated must be less than 2**15
+ * TODO get around the BS alignment bits from the silly pointer variable causing
+ * extra empty padding
+ */
+template <size_t size, typename T = uint16_t>
+class localCircularArenaAllocator : public AllocatorInterface {
+ private:
+  // Compile time check to make sure that the user is using a T large enough to
+  // address size
+  static_assert(size < meta_addressable_space<T>(),
+                "[ERROR](localCircularArenaAllocator) T not large enoughn to "
+                "address size in Arena. Attempted to create Arena with (size, "
+                "T) mismatch, try increasing T to uint32_t");
+
+  class MetaHeader {
+   public:
+    T meta_data;
+    Handle* hndl;
+    uint8_t* _d;
+
+   public:
+    MetaHeader()
+        : meta_data(BLOCK_INACTIVE & BLOCK_ZERO_LENGTH),
+          hndl(nullptr),
+          _d(nullptr) {}
+    MetaHeader(T sz)
+        : meta_data(BLOCK_ACTIVE | sz), hndl(nullptr), _d(nullptr) {}
+    MetaHeader(T sz, uint8_t* d)
+        : meta_data(BLOCK_ACTIVE | sz), hndl(nullptr), _d(d) {}
+    void set_active() { meta_data |= BLOCK_ACTIVE; }
+    void set_inactive() { meta_data &= BLOCK_INACTIVE; }
+    void set_hndl(Handle* handle) { hndl = handle; }
+    void set_d(uint8_t* d) { _d = d; }
+    void set_len(T sz) {
+      meta_data &= MSB_SET;  // Clear all size bits
+      meta_data |= (BLOCK_LENGTH_MASK & sz);
+    }
+    T get_len() const { return meta_data & BLOCK_LENGTH_MASK; }
+    bool is_active() const { return (meta_data & MSB_SET) == BLOCK_ACTIVE; }
+    bool is_bound() const { return (hndl != nullptr); }
+    bool has_handle(Handle* target) const {
+      return is_active() && (hndl == target);
+    }
+    bool is_used() const { return is_active() && (get_len() > 0); }
+  };
+
+ private:
+  T capacity;
+  uint8_t _buffer[size];
+  uint8_t* cursor;
+  std::vector<MetaHeader> _headers;
+  ;
+
+ private:
+  // Return the amount of free space at the tail
+  T tail_capacity(){};
+
+  size_t find_header_associated_w_ptr(void* ptr) const {
+    size_t i = 0;
+    for (auto hdr_i = _headers.begin(); hdr_i != _headers.end(); hdr_i++) {
+      if (hdr_i->_d == ptr) return i;
+    }
+    return i;
+  }
```

Comment:
Looks like `i` is 0 all the time?
So this function will return 0 all the time.

Reply:
You're right this needs to be patched